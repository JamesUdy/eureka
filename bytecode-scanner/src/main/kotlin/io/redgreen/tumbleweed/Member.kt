package io.redgreen.tumbleweed

sealed interface Member {
  val signature: Signature
}
