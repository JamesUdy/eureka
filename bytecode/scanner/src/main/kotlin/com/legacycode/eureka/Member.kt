package com.legacycode.eureka

sealed interface Member {
  val name: String
  val signature: Signature
  val owner: QualifiedType
}
