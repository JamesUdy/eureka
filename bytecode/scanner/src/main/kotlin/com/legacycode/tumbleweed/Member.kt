package com.legacycode.tumbleweed

sealed interface Member {
  val name: String
  val signature: Signature
  val owner: QualifiedType
}
