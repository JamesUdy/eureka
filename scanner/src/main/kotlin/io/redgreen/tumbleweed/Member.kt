package io.redgreen.tumbleweed

interface Member {
  val signature: String
}
