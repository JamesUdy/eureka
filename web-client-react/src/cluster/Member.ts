export class Member {
  constructor(public nodeId: string) {
    // empty
  }

  isMethod() {
    return this.nodeId.endsWith(')');
  }
}
