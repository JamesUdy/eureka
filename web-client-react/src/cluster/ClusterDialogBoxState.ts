import { Member } from './Member';

class SelectionModel {
  private constructor(
    public readonly searchTerm: string,
    public readonly focusedMember: Member | null,
    public readonly filteredMembers: Member[],
  ) {
    // empty
  }

  static DEFAULT = new SelectionModel('', null, []);

  isSearchTermEmpty(): boolean {
    return this.sanitizeSearchTerm(this.searchTerm).length === 0;
  }

  sanitizeSearchTerm(searchTerm: string): string {
    return searchTerm.trim();
  }

  focusNextMember(): SelectionModel {
    const currentFocusedMemberIndex = this.filteredMembers.indexOf(this.focusedMember!);
    const nextFocusedMemberIndex =
      currentFocusedMemberIndex + 1 === this.filteredMembers.length ? 0 : currentFocusedMemberIndex + 1;
    const nextFocusedMember = this.filteredMembers[nextFocusedMemberIndex];

    return new SelectionModel(this.searchTerm, nextFocusedMember, this.filteredMembers);
  }

  focusPreviousMember(): SelectionModel {
    const currentFocusedMemberIndex = this.filteredMembers.indexOf(this.focusedMember!);
    const previousFocusedMemberIndex =
      currentFocusedMemberIndex - 1 < 0 ? this.filteredMembers.length - 1 : currentFocusedMemberIndex - 1;
    const previousFocusedMember = this.filteredMembers[previousFocusedMemberIndex];

    return new SelectionModel(this.searchTerm, previousFocusedMember, this.filteredMembers);
  }

  search(searchTerm: string, members: Member[]): SelectionModel {
    const trimmedSearchTerm = this.sanitizeSearchTerm(searchTerm);

    let filteredMembers: Member[];
    if (trimmedSearchTerm.length > 2) {
      filteredMembers = members.filter((member) =>
        member.nodeId.toLowerCase().includes(trimmedSearchTerm.toLowerCase()),
      );
    } else {
      filteredMembers = [];
    }

    let focusedMember: Member | null = null;
    if (trimmedSearchTerm.length > 0 && filteredMembers.length > 0) {
      focusedMember = filteredMembers[0];
    }
    return new SelectionModel(trimmedSearchTerm, focusedMember, filteredMembers);
  }
}

class ClusterDialogBoxState {
  constructor(
    public readonly members: Member[],
    public readonly startNode: Member | null,
    public readonly startNodeSelectionModel: SelectionModel,
  ) {
    // empty
  }

  static initialState(members: Member[]): ClusterDialogBoxState {
    return new ClusterDialogBoxState(members, null, SelectionModel.DEFAULT);
  }

  search(searchTerm: string): ClusterDialogBoxState {
    const selectionModel = this.startNodeSelectionModel.search(searchTerm, this.members);
    return new ClusterDialogBoxState(this.members, this.startNode, selectionModel);
  }

  isSearchTermEmpty(): boolean {
    return this.startNodeSelectionModel.isSearchTermEmpty();
  }

  focusNextMember(): ClusterDialogBoxState {
    return this.changeFocus(this.startNodeSelectionModel.focusNextMember());
  }

  focusPreviousMember(): ClusterDialogBoxState {
    return this.changeFocus(this.startNodeSelectionModel.focusPreviousMember());
  }

  selectStartNode(): ClusterDialogBoxState {
    return new ClusterDialogBoxState(
      this.members,
      this.startNodeSelectionModel.focusedMember,
      this.startNodeSelectionModel,
    );
  }

  deselectStartNode(): ClusterDialogBoxState {
    return new ClusterDialogBoxState(this.members, null, this.startNodeSelectionModel);
  }

  private changeFocus(selectionModel: SelectionModel) {
    return new ClusterDialogBoxState(this.members, this.startNode, selectionModel);
  }
}

export default ClusterDialogBoxState;
