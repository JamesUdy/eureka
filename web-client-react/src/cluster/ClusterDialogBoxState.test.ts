import ClusterDialogBoxState from './ClusterDialogBoxState';
import { Member } from './Member';

describe('Cluster dialog box state', () => {
  let members: Member[];
  let initialState: ClusterDialogBoxState;

  beforeEach(() => {
    members = [
      new Member('List chrome'),
      new Member('void onCreate()'),
      new Member('void onPause()'),
      new Member('void onResume()'),
    ];
    initialState = ClusterDialogBoxState.initialState(members);
  });

  it('should have an initial state', () => {
    expect(initialState).toMatchSnapshot();
  });

  describe('search functionality', () => {
    it('should not filter members when the search term is empty', () => {
      // when
      const searchState = initialState.search('');

      // then
      expect(searchState).toMatchSnapshot();
    });

    it('should not filter members when the search term only contains spaces', () => {
      // when
      const searchState = initialState.search('      ');

      // then
      expect(searchState).toMatchSnapshot();
      expect(searchState.isSearchTermEmpty()).toBe(true);
    });

    it('should not filter members when the search term has < 2 characters', () => {
      // when
      const searchState = initialState.search('c');

      // then
      expect(searchState).toMatchSnapshot();
    });

    it('should filter members based on the search term', () => {
      // when
      const searchResultState = initialState.search('chr');

      // then
      expect(searchResultState).toMatchSnapshot();
    });

    it('should filter members based on the search term but ignore whitespaces', () => {
      // when
      const searchResultState = initialState.search('  chr  ');

      // then
      expect(searchResultState).toMatchSnapshot();
    });

    it('should not have filtered results for a non-matching search term', () => {
      // when
      const searchResultState = initialState.search('hello');

      // then
      expect(searchResultState).toMatchSnapshot();
      expect(searchResultState.isSearchTermEmpty()).toBe(false);
    });
  });

  describe('focused member functionality', () => {
    let searchResultState: ClusterDialogBoxState;

    beforeEach(() => {
      searchResultState = initialState.search('void');
    });

    it('should select the first matching member and mark it as focused from the search result', () => {
      expect(searchResultState).toMatchSnapshot();
      expect(searchResultState.focusedMember).toEqual(new Member('void onCreate()'));
    });
  });
});
