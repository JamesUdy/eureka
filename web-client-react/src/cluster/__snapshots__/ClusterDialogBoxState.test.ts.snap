// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cluster dialog box state focused member functionality should select the first matching member and mark it as focused from the search result 1`] = `
ClusterDialogBoxState {
  "filteredMembers": Array [
    Member {
      "nodeId": "void onCreate()",
    },
    Member {
      "nodeId": "void onPause()",
    },
    Member {
      "nodeId": "void onResume()",
    },
  ],
  "focusedMember": Member {
    "nodeId": "void onCreate()",
  },
  "members": Array [
    Member {
      "nodeId": "List chrome",
    },
    Member {
      "nodeId": "void onCreate()",
    },
    Member {
      "nodeId": "void onPause()",
    },
    Member {
      "nodeId": "void onResume()",
    },
  ],
  "searchTerm": "void",
}
`;

exports[`Cluster dialog box state search functionality should filter members based on the search term 1`] = `
ClusterDialogBoxState {
  "filteredMembers": Array [
    Member {
      "nodeId": "List chrome",
    },
  ],
  "focusedMember": Member {
    "nodeId": "List chrome",
  },
  "members": Array [
    Member {
      "nodeId": "List chrome",
    },
    Member {
      "nodeId": "void onCreate()",
    },
    Member {
      "nodeId": "void onPause()",
    },
    Member {
      "nodeId": "void onResume()",
    },
  ],
  "searchTerm": "chr",
}
`;

exports[`Cluster dialog box state search functionality should filter members based on the search term but ignore whitespaces 1`] = `
ClusterDialogBoxState {
  "filteredMembers": Array [
    Member {
      "nodeId": "List chrome",
    },
  ],
  "focusedMember": Member {
    "nodeId": "List chrome",
  },
  "members": Array [
    Member {
      "nodeId": "List chrome",
    },
    Member {
      "nodeId": "void onCreate()",
    },
    Member {
      "nodeId": "void onPause()",
    },
    Member {
      "nodeId": "void onResume()",
    },
  ],
  "searchTerm": "chr",
}
`;

exports[`Cluster dialog box state search functionality should not filter members when the search term has < 2 characters 1`] = `
ClusterDialogBoxState {
  "filteredMembers": Array [],
  "focusedMember": null,
  "members": Array [
    Member {
      "nodeId": "List chrome",
    },
    Member {
      "nodeId": "void onCreate()",
    },
    Member {
      "nodeId": "void onPause()",
    },
    Member {
      "nodeId": "void onResume()",
    },
  ],
  "searchTerm": "c",
}
`;

exports[`Cluster dialog box state search functionality should not filter members when the search term is empty 1`] = `
ClusterDialogBoxState {
  "filteredMembers": Array [],
  "focusedMember": null,
  "members": Array [
    Member {
      "nodeId": "List chrome",
    },
    Member {
      "nodeId": "void onCreate()",
    },
    Member {
      "nodeId": "void onPause()",
    },
    Member {
      "nodeId": "void onResume()",
    },
  ],
  "searchTerm": "",
}
`;

exports[`Cluster dialog box state search functionality should not filter members when the search term only contains spaces 1`] = `
ClusterDialogBoxState {
  "filteredMembers": Array [],
  "focusedMember": null,
  "members": Array [
    Member {
      "nodeId": "List chrome",
    },
    Member {
      "nodeId": "void onCreate()",
    },
    Member {
      "nodeId": "void onPause()",
    },
    Member {
      "nodeId": "void onResume()",
    },
  ],
  "searchTerm": "",
}
`;

exports[`Cluster dialog box state search functionality should not have filtered results for a non-matching search term 1`] = `
ClusterDialogBoxState {
  "filteredMembers": Array [],
  "focusedMember": null,
  "members": Array [
    Member {
      "nodeId": "List chrome",
    },
    Member {
      "nodeId": "void onCreate()",
    },
    Member {
      "nodeId": "void onPause()",
    },
    Member {
      "nodeId": "void onResume()",
    },
  ],
  "searchTerm": "hello",
}
`;

exports[`Cluster dialog box state should have an initial state 1`] = `
ClusterDialogBoxState {
  "filteredMembers": Array [],
  "focusedMember": null,
  "members": Array [
    Member {
      "nodeId": "List chrome",
    },
    Member {
      "nodeId": "void onCreate()",
    },
    Member {
      "nodeId": "void onPause()",
    },
    Member {
      "nodeId": "void onResume()",
    },
  ],
  "searchTerm": "",
}
`;
